### PR TITLE
[PS-1471] Create Allocation Free `EncryptedStringAttribute` validation

### PR DESCRIPTION
**Pull request by justindbaur**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [x] Other
```

## Objective

We validate the shape of incoming encrypted strings but we don't actually do anything with the data so every bit of data that is allocated for that adds to the heap that then has to be deallocated later. 

| Method  |                Branch |      EncryptedString |        Mean |     Error |    StdDev |      Gen0 |   Gen1 | Allocated |
| ------- | --------------------: | -------------------- | ----------: | --------: | --------: | --------: | ------:| --------: |
| IsValid | encrypted-string-perf | 2.Kl(...)uXg= [5616] |    375.8 ns |   4.47 ns |   3.49 ns |         - |      - |         - |
| IsValid | encrypted-string-perf | 2.QmF(...)ydA== [52] |    137.8 ns |   2.79 ns |   2.87 ns |         - |      - |         - |
| IsValid |                master | 2.Kl(...)uXg= [5616] | 15,731.6 ns | 304.21 ns | 362.14 ns |    4.3640 | 0.2441 |   27425 B |
| IsValid |                master | 2.QmF(...)ydA== [52] |    314.5 ns |   6.24 ns |   6.93 ns |    0.0839 |      - |     528 B |

NOTE: It's worth stating that a this attribute is called a few times in each request that uses them. For login it's used probably 4 times (Name, Username, Password, Uri) so the savings would be 4x at least what is shown in low `[52]` character input.

## Code changes

* **bitwarden-server.sln:** Added perf project so we can have a single place to add benchmarks.
* **perf/MicroBenchmarks/Core/EncryptedStringAttributeTests.cs:** A benchmark that tests the current implementation of `IsValid` with two parameters, one that is a general "password" and one that is 5 paragraphs of lorem ipsum.
* **src/Core/Utilities/EncryptedStringAttribute.cs:** Refactor of `IsValid` to be quicker and to have less allocations.
* **src/Core/Utilities/SpanExtensions.cs:** Add Span extension for splitting a ReadOnlySpan<char> by a certain character and returning whether it was successful and using `out` parameters to return both a span before and after the split character.
* **test/Core.Test/Utilities/EncryptedStringAttributeTests.cs:** Added tests that more thourougly test the possible inputs to the validation and it's expected input
  * These were added before any changes to the underlying logic was made and all were successful against that with no uncovered branches (except for one unreachable branch). 

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by MGibson1 — CHANGES_REQUESTED** [review]

File: src/Core/Utilities/SpanExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+﻿namespace Bit.Core.Utilities;
+
+public static class SpanExtensions
+{
+    public static bool TrySplitBy(this ReadOnlySpan<char> input,
+        char splitChar, out ReadOnlySpan<char> chunk, out ReadOnlySpan<char> rest)
+    {
+        var splitIndex = input.IndexOf(splitChar);
+
+        if (splitIndex < 1)
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply src/Core/Utilities/SpanExtensions.cs:10]: That's a good point, I find myself adding `StringSplitOptions.RemoveEmptyEntries` all the time but I should and can change this implementation to consider that a success without hurting the implementation I am using it for.

File: src/Core/Utilities/EncryptedStringAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            }
+
+            if (value is string stringValue)
+            {
+                // Fast path
+                return IsValidCore(stringValue);
+            }
+
+            // Slow path (ish)
+            return IsValidCore(value.ToString());
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply src/Core/Utilities/EncryptedStringAttribute.cs:45]: In the current project there are none, but there would be no compile time issue with adding the attribute to non-string properties and the existing implementation supported this scenario so I decided to as well but would be thrilled to remove it and consider it invalid.

File: src/Core/Utilities/EncryptedStringAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+            var pieces = 1;
+            var findIndex = encryptionPiecesChunk.IndexOf('|');
+
+            while (findIndex != -1)
+            {
+                pieces++;
+                encryptionPiecesChunk = encryptionPiecesChunk[++findIndex..];
+                findIndex = encryptionPiecesChunk.IndexOf('|');
+            }
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply src/Core/Utilities/EncryptedStringAttribute.cs:71]: There currently isn't a built-in one but most likely there will be in [.NET 8](https://github.com/dotnet/runtime/issues/59466). I can make our own helper for this and then replace it with the built in, faster implementation when it's released.

File: src/Core/Utilities/EncryptedStringAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        }
+
+        EncryptionType encryptionType;
+
+        // Using byte here because that is the backing type for EncryptionType
+        if (!byte.TryParse(headerChunk, out var encryptionTypeNumber))
+        {
+            // We can't read the header chunk as a number, this is the slow path
+            if (!Enum.TryParse(headerChunk, out encryptionType))
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply src/Core/Utilities/EncryptedStringAttribute.cs:89]: I would love to remove it but the current implementation supports the name AND the existing tests use the name so I was worried about remove it fully. I know all the strings I have ever seen have the number and that is what I use as the fast path but would be happy to just fail quick too.

File: src/Core/Utilities/EncryptedStringAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        return rest.IndexOf('|') == -1;
+    }
+
+    private static bool IsValidBase64(ReadOnlySpan<char> input)
+    {
+        byte[]? pooledChunks = null;
+
+        // Ref: https://vcsjones.dev/stackalloc/
+        var byteBuffer = input.Length > 128
+            ? (pooledChunks = ArrayPool<byte>.Shared.Rent(input.Length * 2))
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply src/Core/Utilities/EncryptedStringAttribute.cs:160]: Not by default but they have an overload on `Return` for `bool clearArray` that we can use.

File: src/Core/Utilities/EncryptedStringAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            : stackalloc byte[256];
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply src/Core/Utilities/EncryptedStringAttribute.cs:161]: This is my least favorite too. I don't even want to write the base64 anywhere and just do in-memory validation that it's valid base64 but it's not built in. 256 comes from char being 2 bytes but yes you are absolutely correct that we should need a smaller byte buffer to store the validated base64. I will rewrite it to take that into account and then more of the paths will be able to allocated on the stack!

File: src/Core/Utilities/EncryptedStringAttribute.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MGibson1) [reply src/Core/Utilities/EncryptedStringAttribute.cs:79]: oh, my mistake. I thought we were currently failing in the event we got the enum name. We should keep this. It should always go through the fast path anyways 👍

**Review by MGibson1 — APPROVED** [review]
This looks safe to me

**Review by MGibson1 — COMMENTED** [review]
why'd this turn into a 13k change PR?

**Comment by justindbaur** [comment]
> 

@MGibson1 I think something weird happened with the rebase. I reverted that commit and did it again.

**Review by MGibson1 — APPROVED** [review]

**Comment by MGibson1** [comment]
Thanks for fixing. still a huge changeset, but not it's entirely due to the package.lock file for the new csproj

**Pull request merged**